collect_app/src/main/java/org/odk/collect/android/external/FormUriActivity.kt

```diff
@@ -317,6 +327,19 @@ private class FormUriViewModel(
         }
     }
 
+    private suspend fun assertFormsUpdateNotInProgress(): String? {
+        val projectId = projectsDataService.getCurrentProject().uuid
+        val formsLock = changeLockProvider.create(projectId).formsLock
+        val isLocAcquired = withContext(Dispatchers.Main) {
```

Comment:
OK so I think we want to implement `ChangeLock` like this:

```kotlin
class ThreadSafeBooleanChangeLock : ChangeLock {

    private var locked = false

    override fun <T> withLock(function: Function<Boolean, T>): T {
        val acquired = tryLock()

        try {
            return function.apply(acquired)
        } finally {
            unlock()
        }
    }

    override fun tryLock(): Boolean {
        return synchronized(this) {
            if (!locked) {
                locked = true
                true
            } else {
                false
            }
        }
    }

    override fun unlock() {
        synchronized(this) {
            locked = false
        }
    }
}
```

Basically, with `ChangeLock` we're trying to signal when it's safe to make changes to something. `ReentrantChangeLock` does that, but has the added complexity of adding lock "ownership" (a pretty common need in a lot of use cases). Before, that works as we only ever locked a `ChangeLock` for the execution of one function and that would always happen with a single thread that could safely own the lock. Now we want to lock the `ChangeLock` while a user interacts with a certain portion of the app, so it's not appropriate for a thread to "acquire" the lock. What we want instead is a simpler lock that's just locked/unlocked (like `BooleanChangeLock`) but that can be safely accessed by multiple threads concurrently. This achieves that by synchronizing access and modification of the `locked` field, but still allows unsynchronized execution of code trying to lock in `withLock`.

@lognaturel it'd be good to get your eyes on this as well as concurrency is hard and the more people's brains the better!

Reply:
> we could definitely run into problems granting the whole UI thread the lock here.

Could you please list some examples? I can't seem to convince myself that there are likely problems!

I agree that the high-level goal here is not thread-specific: there are clearly multiple threads involved in form entry and the goal is to prevent form download when any form entry related behavior is ongoing.

That said, the UI thread "orchestrates" all of the behavior related to form entry and all of form entry is synchronous and user-facing so it seems like having the UI thread hold the lock would have the desired behavior. Any other thread attempting to do form-related work would be unable to acquire the lock. The UI thread could re-acquire the lock but I believe that form entry is the only task that happens on the UI thread which needs the form lock. Regardless, there's no way for users to go from form entry to another ui-based task without exiting form entry. `ReentrantLock.tryLock` is designed for concurrency so there shouldn't be any issue there.

The biggest risk in my mind is that we have to be quite confident that we release the lock in every possible way there is for form entry to end. I don't think the proposed alternative changes this in any way.

> added complexity of adding lock "ownership" (a pretty common need in a lot of use cases)

This is mostly to avoid deadlock, right?

> We'll also need to remove any superfluous lock checks we have if we go with an implementation like I've suggested above

Deadlock is scary! Would be really great not to have to worry about it.

Reply:
> Could you please list some examples? I can't seem to convince myself that there are likely problems!

Off the top of my head, one weirdness you can see from the current implementation is that we end up having to schedule code to run on the UI thread from a background thread so that the UI thread gets the lock rather than the background thread actually dealing with opening the form. Another example would be that form saving heavily interacts with the forms DB, but wouldn't technically have the lock. That's OK because it doesn't check the lock, but it emphasises that having the lock owned by a thread is pretty pointless. I'm mostly worried about us tangling ourselves into a race condition/deadlock that we're not expecting because we're using thread based locking in a situation that it's not really suited to.

> This is mostly to avoid deadlock, right?

I don't think so. Locks granting ownership on a ("native") thread by thread basis is just their standard use case right? Preventing multiple threads from access a resource at the same time. As you hypothesise, I think whether we use a standard lock that the UI thread owns or move to a more basic "flag" like implementation (like I suggested above) the risk of a deadlock is the same as the "unlock" is not ensured by something like a `finally`.

Reply:
> form saving heavily interacts with the forms DB, but wouldn't technically have the lock

I see, I agree that doesn't feel right. Form saving is initiated by and blocks the UI thread but it's not actually performed by the UI thread.

> Locks granting ownership on a ("native") thread by thread basis is just their standard use case right? Preventing multiple threads from access a resource at the same time.

Err, not sure exactly what I meant there. Maybe more the reentrant part -- that the same thread can get the same lock again? When you said "superfluous lock checks" you mean nested checks that the reentrant lock can handle, right?

Can we use a binary semaphore -- [Semaphore](https://developer.android.com/reference/java/util/concurrent/Semaphore) initialized with 1 permit? That feels like it matches the semantics we're aiming for.

Would using acquire` instead of `tryAcquire` in the form update cases let those updates go through as soon as form entry completes? That seems like it could be better than skipping the current update.

Reply:
> Can we use a binary semaphore -- [Semaphore](https://developer.android.com/reference/java/util/concurrent/Semaphore) initialized with 1 permit? That feels like it matches the semantics we're aiming for and then maybe we don't have to introduce our own language.

I think so? When I'd looked at this on Thursday, I'd been thinking that a semaphore requires the same thread to lock/unlock (acquire/release in a semaphore's case), but that should not be true. I'm guessing that was the beginnings of the cold kicking in 🤣. I think for the moment we should keep the naming the same to limit the changes we need to make in this PR and go with the implementation suggested in https://github.com/getodk/collect/pull/6440#discussion_r1804599427 (which I'm pretty sure is just a binary sempahore as @lognaturel points out). We can follow up with another PR that reworks the language to make it clearer.

> Would using acquire instead of tryAcquire in the form update cases let those updates go through as soon as form entry completes? That seems like it could be better than skipping the current update.

As far as I'm aware that would involve blocking threads in the background which Android will not like. I think the "try and schedule retry" approach is correct for the context as it allows threads to be returned to their pool.

Reply:
> I think for the moment we should keep the naming the same to limit the changes we need to make in this PR and go with the implementation suggested in https://github.com/getodk/collect/pull/6440#discussion_r1804599427

Done.